stdlib/public/core/Process.swift

```diff
@@ -12,44 +12,50 @@
 
 import SwiftShims
 
-internal class _Box<Wrapped> {
-  internal var _value: Wrapped
-  internal init(_ value: Wrapped) { self._value = value }
-}
-
 /// Command-line arguments for the current process.
 public enum Process {
   /// Return an array of string containing the list of command-line arguments
   /// with which the current process was invoked.
   internal static func _computeArguments() -> [String] {
-    var result: [String] = []
-    let argv = unsafeArgv
-    for i in 0..<Int(argc) {
-      let arg = argv[i]!
-      let converted = String(cString: arg)
-      result.append(converted)
-    }
-    return result 
+    let argv = self.unsafeArgv
+    return (0..<Int(argc)).map { String(cString: argv[$0]!) }
   }
 
   @_versioned
+  /// The backing static variable for argument count may come either from the
+  /// entry point or it may need to be computed e.g. if we're in the REPL.
   internal static var _argc: CInt = CInt()
 
   @_versioned
   internal static var _unsafeArgv:
-    UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?
-    = nil
+    /// The backing static variable for arguments may come either from the
+    /// entry point or it may need to be computed e.g. if we're in the REPL.
+    ///
+    /// Care must be taken to ensure that `_swift_stdlib_getUnsafeArgvArgc` is
+    /// not invoked more times than is necessary (at most once).
+    UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>? = nil
 
   /// Access to the raw argc value from C.
   public static var argc: CInt {
+    _ = Process.unsafeArgv // Force strict evaulation of _unsafeArgv.
     return _argc
   }
 
   /// Access to the raw argv value from C. Accessing the argument vector
   /// through this pointer is unsafe.
   public static var unsafeArgv:
     UnsafeMutablePointer<UnsafeMutablePointer<Int8>?> {
-    return _unsafeArgv!
+    let argumentsPtr = UnsafeMutablePointer<AnyObject?>(
+      Builtin.addressof(&_swift_stdlib_ProcessUnsafeArgv))
+
+    let newArgs = _unsafeArgv ?? _swift_stdlib_getUnsafeArgvArgc(&_argc)
+    if _stdlib_atomicCompareExchangeStrongPtr(
+      object: argumentsPtr, expected: nil, desired: newArgs) {
+      // On platforms where `_swift_stdlib_getUnsafeArgvArgc` performs a call to
+      // malloc and loses the race, we need to free the extra pointer.
+      _swift_stdlib_freeUnsafeArgvArgc(newArgs, _argc)
```

Comment:
This isn't correct if we lose the race but `_unsafeArgv` *was* set. I'm still in favor of my other approach that just uses static variables.